broker/lib/fabrik/DirectorInstance.js

```diff
@@ -240,21 +243,60 @@ class DirectorInstance extends BaseInstance {
         );
     }
     logger.info('Fetching state of last operation', operation);
-    return Promise
-      .try(() => {
-        assert.ok(operation.task_id, 'Operation must have the property \'task_id\'');
-        return this.manager.getTask(operation.task_id);
-      })
-      .catchThrow(NotFound, new ServiceInstanceNotFound(this.guid))
-      .then(task => {
-        assert.ok(_.endsWith(task.deployment, this.guid), `Deployment '${task.deployment}' must end with '${this.guid}'`);
-        this.networkSegmentIndex = this.manager.getNetworkSegmentIndex(task.deployment);
-        this.setOperationState(operation, task);
-        if (operation.state !== 'in progress') {
-          return this.finalize(operation);
+    const instanceId = this.guid;
+
+    const getBoshTaskOperation = (operation, taskId, removeFromCache) => {
+      return Promise
+        .try(() => {
+          assert.ok(taskId, 'Task ID must be available');
+          return this.manager.getTask(taskId);
+        })
+        .catchThrow(NotFound, new ServiceInstanceNotFound(instanceId))
+        .then(task => {
+          assert.ok(_.endsWith(task.deployment, this.guid), `Deployment '${task.deployment}' must end with '${this.guid}'`);
+          this.networkSegmentIndex = this.manager.getNetworkSegmentIndex(task.deployment);
+          this.setOperationState(operation, task);
+          if (operation.state !== 'in progress') {
+            return Promise.try(() => {
+                if (removeFromCache) {
+                  return utils
+                    .retry(tries => {
+                      logger.info(`+-> ${ordinals[tries]} attempt to delete cached bosh task for '${instanceId}'...`);
+                      return this.manager.removeCachedTask(instanceId);
+                    }, {
+                      maxAttempts: 5,
+                      minDelay: 1000
+                    })
+                    .catch(err => {
+                      logger.error('Error removing task from ETCD Cache after multiple attempts', err);
+                      throw err;
+                    });
+                }
+              })
+              .then(() => this.finalize(operation));
+          }
+        })
+        .return(operation);
+    };
+
+    if (operation.task_id) {
+      return getBoshTaskOperation(operation, operation.task_id, false);
+    } else {
+      return Promise.try(() => {
+        return this.manager.getCurrentOperationState(this.guid);
+      }).then(state => {
+        const isCached = state.cached;
+        const taskId = state.task_id;
+        if (isCached) {
+          return _.assign(operation, {
+            description: `${_.capitalize(operation.type)} deployment is still in progress`,
+            state: 'in progress'
+          });
+        } else {
+          return getBoshTaskOperation(operation, taskId, true);
```

Comment:
There are three cases here:
- If the bosh rate limit feature is not enabled, only the task_id is available [reverts to current flow]
- If the bosh rate limit feature is enabled and there were empty slots in BOSH, only the task_id is available [reverts to current flow]
- If the bosh rate limit feature is enabled and there were no empty slots in BOSH, the operation is queued in Etcd and the director task poller runs every minute checking against BOSH
   - In between, the last operation request from CC could arrive at the broker. In this state, the broker checks against two etcd keys [mentioned below] and if it finds that a task_id has already been generated, it checks against BOSH for the status [current flow]. If it finds that a task_id has not been generated yet and the entry is still in ETCD, it returns back `in_progress`
   - As soon as a slot opens up in BOSH, the poller attempts the operation against BOSH and removes the initial entry from etcd. It now gets a task ID from BOSH and pushes it into another key in ETCD against the instance ID

So, if the task is not queued on BOSH yet, the task ID will not be available but in this case, the `getBoshTaskStatus` would be invoked and it would fail there anyway since an `assert` is already there